{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PyRosetta-4 2020 [Rosetta PyRosetta4.conda.linux.cxx11thread.serialization.CentOS.python37.Release 2020.25+release.d2d9f90b8cbcacfd7a1f69aefa5de610b100e8a9 2020-06-19T14:33:13] retrieved from: http://www.pyrosetta.org\n",
      "(C) Copyright Rosetta Commons Member Institutions. Created in JHU by Sergey Lyskov and PyRosetta Team.\n",
      "\u001b[0mcore.init: {0} \u001b[0mChecking for fconfig files in pwd and ./rosetta/flags\n",
      "\u001b[0mcore.init: {0} \u001b[0mRosetta version: PyRosetta4.conda.linux.cxx11thread.serialization.CentOS.python37.Release r258 2020.25+release.d2d9f90 d2d9f90b8cbcacfd7a1f69aefa5de610b100e8a9 http://www.pyrosetta.org 2020-06-19T14:33:13\n",
      "\u001b[0mcore.init: {0} \u001b[0mcommand: PyRosetta -ex1 -ex2aro -database /home/iwe2/anaconda3/envs/pyrosetta_env/lib/python3.7/site-packages/pyrosetta/database\n",
      "\u001b[0mbasic.random.init_random_generator: {0} \u001b[0m'RNG device' seed mode, using '/dev/urandom', seed=-187899166 seed_offset=0 real_seed=-187899166 thread_index=0\n",
      "\u001b[0mbasic.random.init_random_generator: {0} \u001b[0mRandomGenerator:init: Normal mode, seed=-187899166 RG_type=mt19937\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pyrosetta import *\n",
    "init()\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from mpl_toolkits import mplot3d\n",
    "from scipy.spatial import *\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import *\n",
    "\n",
    "import os\n",
    "os.system('ls *.pdb > pdb.list')\n",
    "#import pyrosetta for distance calculatingand Biopython for alignment\n",
    "#generate a list with all pdbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['10_S_0001.pdb', '1_S_0001.pdb', '2_S_0001.pdb', '3_S_0001.pdb', '4_S_0001.pdb', '5_S_0001.pdb', '6_S_0001.pdb', '7_S_0001.pdb', '8_S_0001.pdb', '9_S_0001.pdb']\n"
     ]
    }
   ],
   "source": [
    "def readin_files(file):\n",
    "    data_list = []\n",
    "    with open(file) as open_file:\n",
    "        for line in open_file:\n",
    "            data_list.append(line[:-1])\n",
    "    return data_list\n",
    "#defines the reading of files\n",
    "\n",
    "model = readin_files('pdb.list')\n",
    "assignment = readin_files('assignment.list')\n",
    "volume_residues = readin_files('selected_res_vol.list')\n",
    "print(model)\n",
    "#read in all the necessary files and assignments\n",
    "\n",
    "#####################################\n",
    "#Define your target name from before#\n",
    "#####################################\n",
    "target_name = 'ADRB1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '10_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "10_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '1_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "1_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '2_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '3_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "3_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '4_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "4_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '5_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "5_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '6_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "6_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '7_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "7_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '8_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "8_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "> ADRB1\n",
      "\u001b[0mcore.import_pose.import_pose: {0} \u001b[0mFile '9_S_0001.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mFound disulfide between residues 83 168\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYS\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: {0} \u001b[0mcurrent variant for 168 CYD\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "9_S_0001.pdb\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "error = []\n",
    "for structure in model:\n",
    "    try:\n",
    "        pdb_file = structure.split()\n",
    "        for assign in assignment:\n",
    "            assign = assign.split('\\t')\n",
    "            if target_name == assign[0]:\n",
    "                bw_assign = \"> \" + assign[2]\n",
    "            if target_name == assign[2]:\n",
    "                bw_assign = \"> \" + assign[2]\n",
    "            \n",
    "    #Here the script tries to find the model in the assignment list, so to define bw_assign for the later BW file\n",
    "        print(bw_assign)\n",
    "        pose = pose_from_pdb(structure)\n",
    "        fasta_model = pose.sequence()\n",
    "    #The typical Pyrosetta start to read in the pdb and work with it\n",
    "\n",
    "        file = open('all_gpcrdb_num.data', 'r')\n",
    "        lines = file.readlines()\n",
    "        for i in range (0, len(lines)):\n",
    "            if bw_assign in lines[i]:\n",
    "                position = i\n",
    "        print(bw_assign)\n",
    "    #The script gets the information, where in the BW list to find the needed residues\n",
    "\n",
    "        fasta_bw = ''\n",
    "        fasta = lines[position + 1]\n",
    "        fasta = fasta.split('\\t')\n",
    "        fasta[-1] = fasta[-1][:-1]\n",
    "        bw = lines[position + 2]\n",
    "        bw = bw.split('\\t')\n",
    "        bw[-1] = bw[-1][:-1]\n",
    "        for ele in fasta:\n",
    "            fasta_bw = fasta_bw + ele[0]\n",
    "        #From the BW all data list the sequence for BW numbering is taken, \n",
    "        #defined and a list with the names provided\n",
    "\n",
    "        alignments = pairwise2.align.globalxx(fasta_model, fasta_bw)\n",
    "        #use of Biophython lib and the module pairwise 2\n",
    "        aligned_bw = alignments[0][1]\n",
    "        aligned_pdb = alignments[0][0]\n",
    "        #the alignment will put '-' in spaces were there are differences\n",
    "\n",
    "        offset_bw = 0\n",
    "        offset_pdb = 0\n",
    "        if aligned_pdb[0] == '-':\n",
    "            for residue in aligned_pdb:\n",
    "                if residue != '-':\n",
    "                    break\n",
    "                else:\n",
    "                    offset_pdb += 1\n",
    "        else:\n",
    "            for residue in aligned_bw:\n",
    "                if residue != '-':\n",
    "                    break\n",
    "                else:\n",
    "                    offset_bw += 1\n",
    "        #need to define an offset value (bw sequence starts earlier or later)\n",
    "\n",
    "        bw_for_alignment = []           \n",
    "        #print(offset_bw, offset_pdb)\n",
    "\n",
    "\n",
    "        offset_aligned_in_fasta = 0\n",
    "        offset_bw_in_fasta = 0\n",
    "\n",
    "        bw_numbering_final = []\n",
    "        if offset_bw > offset_pdb: \n",
    "            for i in range(offset_bw, len(fasta_model)):\n",
    "                if aligned_bw[i] == '-':\n",
    "                    offset_aligned_in_fasta += 1 \n",
    "                else:\n",
    "                    try:\n",
    "                        if aligned_pdb[i] == '-':\n",
    "                            offset_bw_in_fasta += 1 \n",
    "                        else:\n",
    "                            bw_numbering_final.append((aligned_pdb[i], i + 1 - offset_bw_in_fasta, \n",
    "                                                       bw[i - offset_bw - offset_aligned_in_fasta]))\n",
    "                    except:\n",
    "                        pass\n",
    "    #Here happens the magic \n",
    "    #We need two further offset values, one offset_aligned_in_fasta for the bw numbering and \n",
    "    #offset_bw_in_fasta for when in the crystal structure we have\n",
    "    #another residue (complicated, but the assignment will make from ex. VASQAS BW and VASNAS determined structure \n",
    "    #-> VAS-QAS and VASN-AS)\n",
    "\n",
    "        else:\n",
    "            for i in range(offset_pdb, len(fasta_model)):\n",
    "                if aligned_bw[i] == '-':\n",
    "                    offset_aligned_in_fasta  += 1  \n",
    "                else:\n",
    "                    try:                    \n",
    "                        if aligned_pdb[i] == '-':\n",
    "                            offset_bw_in_fasta += 1 \n",
    "                        else:\n",
    "                            bw_numbering_final.append((aligned_pdb[i], (i + 1 - offset_pdb - offset_bw_in_fasta), \n",
    "                                                       bw[i - offset_aligned_in_fasta]))\n",
    "                    except:\n",
    "                        pass\n",
    "        #prints AA, number in Rosetta and BW Numbering        \n",
    "\n",
    "\n",
    "        new_file = f\"{structure}.xyz_vol\"\n",
    "        f = open(new_file, \"w\")\n",
    "        relevant = []\n",
    "        for ele in volume_residues:\n",
    "            for row in bw_numbering_final:\n",
    "                if ele in row:\n",
    "                    relevant.append(row)\n",
    "\n",
    "            if relevant[-1][-1] == ele:\n",
    "                pass\n",
    "            else:\n",
    "                relevant.append(\"NA\")\n",
    "        distance = []\n",
    "        x_val = len(relevant)\n",
    "        for i in range(0, (x_val * x_val)):\n",
    "            distance.append(0.0)\n",
    "\n",
    "        res_xyz = []\n",
    "        for line in relevant:\n",
    "            try:\n",
    "        #IMPORTANT!!!! I take chain A, but its also possible that the interactions are from chain B, CHECK THIS!!!\n",
    "                res = pose.residue(int(line[1]))\n",
    "                res_xyz.append(res.xyz(\"CA\"))\n",
    "            except:\n",
    "                res_xyz.append(\"NA\")\n",
    "        for i in range(0, len(res_xyz)):\n",
    "            print(*relevant[i], res_xyz[i], sep='\\t', file=f)\n",
    "        f.close()\n",
    "    except:\n",
    "        print('\\n\\n\\n\\n' + str(structure) + '\\n\\n\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['4BVN.pdb']\n"
     ]
    }
   ],
   "source": [
    "def readin_files(file):\n",
    "    data_list = []\n",
    "    with open(file) as open_file:\n",
    "        for line in open_file:\n",
    "            data_list.append(line[:-1])\n",
    "    return data_list\n",
    "#defines the reading of files\n",
    "\n",
    "model = readin_files('pdb.list')\n",
    "residues = readin_files('selected_res.list')\n",
    "assignment = readin_files('assignment.list')\n",
    "volume_residues = readin_files('selected_res_vol.list')\n",
    "print(model)\n",
    "#read in all the necessary files and assignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n"
     ]
    }
   ],
   "source": [
    "#For calculating the volume of the 10 res\n",
    "xyz = []\n",
    "for structure in model:\n",
    "    xyz.append((structure + \".xyz_vol\"))\n",
    "\n",
    "counter = 0\n",
    "for ele in xyz:\n",
    "    new_file = f\"{ele}.vol\"\n",
    "    f = open(new_file, \"w\")\n",
    "    points = []\n",
    "    file = open(ele, 'r')\n",
    "    try:\n",
    "        for line in file:\n",
    "            line = line.split()\n",
    "            if \"NA\" not in line:\n",
    "                point = line[3:]\n",
    "            points.append(point)\n",
    "        hull = ConvexHull(points)\n",
    "        print(ele, hull.volume, file=f)\n",
    "        f.close()\n",
    "        counter += 1\n",
    "        print(counter)\n",
    "    except:\n",
    "        print(ele, \"NA\", file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-25.571\n",
      "-22.962\n",
      "-17.396\n",
      "-16.433\n",
      "-20.153\n",
      "-20.647\n",
      "-24.016\n",
      "-14.499\n",
      "-17.04\n",
      "-18.626\n",
      "-19.534\n",
      "-22.143\n",
      "-21.007\n",
      "-24.504\n",
      "-16.319\n",
      "-14.711\n",
      "-23.316\n",
      "-26.183\n",
      "-23.37\n",
      "-26.642\n",
      "-23.179\n",
      "-26.646\n",
      "-30.817\n",
      "-31.844\n",
      "-31.317\n",
      "-31.037\n",
      "-33.339\n",
      "-31.026\n",
      "-31.326\n",
      "-31.835\n",
      "-29.497\n",
      "-28.598\n",
      "-28.466\n",
      "-28.462\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "#For calculating the distance matrix\n",
    "xyz = []\n",
    "for structure in model:\n",
    "    xyz.append((structure + \".xyz\"))\n",
    "\n",
    "counter = 0\n",
    "for ele in xyz:\n",
    "    new_file = f\"{ele}.matrix\"\n",
    "    f = open(new_file, \"w\")\n",
    "    res_coordinates = []\n",
    "    bw_number = []\n",
    "    file = open(ele, 'r')\n",
    "    for line in file:\n",
    "        line = line.split()\n",
    "        if \"NA\" not in line:\n",
    "            x = float(line[3])\n",
    "            y = float(line[4])\n",
    "            z = float(line[5])\n",
    "            coordinate = np.array((x,y,z))\n",
    "            res_coordinates.append(coordinate)\n",
    "            bw_number.append(line[2])\n",
    "\n",
    "    for i in range(0, len(res_coordinates)):\n",
    "        for o in range(i+1, len(res_coordinates)):\n",
    "\n",
    "            distance = np.linalg.norm(np.array((res_coordinates[i]) - np.array(res_coordinates[o])))\n",
    "            print(bw_number[i], 'x', bw_number[o], distance, sep='\\t', file=f)\n",
    "\n",
    "    f.close()\n",
    "    counter += 1\n",
    "    print(counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
